### PR TITLE
[Shiny] Resolve<IEnumerable<T>>() or ResolveAll<T>()

### PR DESCRIPTION
**Pull request by dahlbyk**
I initially reported this as https://github.com/shinyorg/shiny/issues/359, but @aritchie added some tests that suggest the issue is specific to Shiny.Prism. Specifically, `Resolve<IEnumerable<T>>()` is returning `null` instead of the expected collection. (I opted not to upgrade to Shiny 1.2, which introduces an equivalent `ResolveAll<T>()` extension.)

It looks like DryIoc can `Resolve<IEnumerable<T>>()` as expected, so the issue seems specific to how Shiny is being set up. Any ideas, @dansiegel?

**Note:** PR is branched from v7.2.0.1054 rather than red `master`.

### CODE REVIEW

**Comment by aritchie** [comment]
Hey @dansiegel - I think GetService should return null when no service is found, GetServices should return an empty enumerable

**Comment by dahlbyk** [comment]
> I think GetService should return null when no service is found, GetServices should return an empty enumerable

Agreed, I was surprised to find different behavior from `ShinyHost.ResolveAll<T>()` (exception) and `ShinyHost.Container.ResolveAll<T>()` (null). Of course, neither behavior is correct but it's also odd that their inconsistent.

**Comment by aritchie** [comment]
@dahlbyk I've resolved the inconsistencies between the two ResolveAll calls in the myget feed.  I've also put safeties on locations where service enumerables should be checking for null

File: tests/Shiny.Prism.Tests/Tests/PrismStartupTests.cs [hunk 1]
```diff
@@ -53,6 +58,27 @@ public void ExpectedTypesAreRegisteredAndResolve(Type serviceType, Type implemen
...
+            });
+
+            Assert.True(PrismContainerExtension.Current.IsRegistered(typeof(IPushDelegate)));
+
+            var fromShiny = ShinyHost.Resolve<IPushDelegate>();
+            Assert.IsType<MockPushDelegate>(fromShiny);
+            Assert.Same(fromShiny, PrismContainerExtension.Current.Resolve(typeof(IPushDelegate)));
+            Assert.Same(fromShiny, ShinyHost.Container.Resolve<IPushDelegate>());
+
+            Assert.Equal(new[] { fromShiny }, PrismContainerExtension.Current.Resolve<IEnumerable<IPushDelegate>>());
```
Comment (dahlbyk) [inline tests/Shiny.Prism.Tests/Tests/PrismStartupTests.cs:76]: Note the test is failing on Line 77. This `Resolve` works as expected, so the issue has to be somewhere in the Shiny integration.

**Comment by dansiegel** [comment]
@aritchie I seem to remember discussing that issue with you when I was originally working on this... anytime that IServiceProvider.Resolve is called it should return null and anytime Prism's IContainerProvider is called it should return an exception... if it's doing anything other than that it's a bug.

**Comment by aritchie** [comment]
@dansiegel I think it is a call to GetServices that is missing.  So singular calls return null, enumerables should return empty.  

```csharp
// I expect null here
var items = container.GetService<IEnumerable<MyType>>();

// I expect an empty enumerable here
var items = container.GetServices<MyType>();
```

I think where this actually becomes a real problem is Shiny now allows multiple delegates of the same type.  I use the service provider with a call to GetServices which is returning null even though an item is registered.  I believe the issue is here: 

https://github.com/dansiegel/Prism.Container.Extensions/blob/master/src/Prism.DryIoc.Extensions/PrismContainerExtension.cs#L307

I haven't debugged this myself, but if I have to guess, the serviceType is coming in as IEnumerable<T> which is indeed not registered in the container directly.

**Comment by dansiegel** [comment]
@dahlbyk you probably will want to update this PR to use the latest release of Shiny

**Comment by aritchie** [comment]
@dansiegel I didn't fix the issue on my side.  I can't really.  In fact, returning an empty enumerable when I get null back for the delegates may actually be masking an issue.  There are delegates/services registered under the contract, but because I'm calling GetServices which calls for an IEnumerable of the delegate, nothing is being returned.

**Comment by dahlbyk** [comment]
> I believe the issue is here:
> 
> https://github.com/dansiegel/Prism.Container.Extensions/blob/06bc814a13c146319270a6a692bce5fe09770ed7/src/Prism.DryIoc.Extensions/PrismContainerExtension.cs#L305-L310
> 
> I haven't debugged this myself, but if I have to guess, the serviceType is coming in as IEnumerable which is indeed not registered in the container directly.

This is correct. I believe I've pushed a reasonable fix.

**Comment by dansiegel** [comment]
@aritchie keep in mind GetServices is an extension method that comes from Microsoft.Extensions.DependencyInjection and not actually on IServiceProvider. I suppose with some effort the PrismServiceProvider could look to see if the requested type was an IEnumerable and create an empty implementation. 

@dahlbyk as I've started looking at this, this PR won't actually solve what it needs to as the only actual code change is to `IsRegistered`. Also there is a flaw in the tests as it's giving us a false positive. I'm going to close this for now while we discuss the right way forward in #124

**Comment by dahlbyk** [comment]
By "false positive" I assume you mean the test implies `ResolveAll()` works for all scenarios, but in fact only returns a collection with the most recently registered service?

**Comment by dansiegel** [comment]
@dahlbyk in my testing it was only returning a single element not all elements that were registered.

**Comment by dahlbyk** [comment]
> @dahlbyk in my testing it was only returning a single element not all elements that were registered.

I believe that, based on https://github.com/dansiegel/Prism.Container.Extensions/issues/124#issuecomment-650963347.

I'm pretty sure we _also_ need the change in this PR, though. Tests for multiple registrations should be added once https://github.com/PrismLibrary/Prism/pull/2143 lands, but with or without that change I believe you'll still run into `IsRegistered(typeof(IEnumerable<IAmRegistered>))` returning `false`. Thus `GetService()` will continue to return `null` instead of the expected collection:

https://github.com/dansiegel/Prism.Container.Extensions/blob/06bc814a13c146319270a6a692bce5fe09770ed7/src/Prism.DryIoc.Extensions/PrismContainerExtension.cs#L305-L310

**Comment by dansiegel** [comment]
@dahlbyk you are right... we can make that change to ensure that IRegistered returns true when IEnumerable<T> is called and it was T that was registered. We'll start with the change in 2143...

File: src/Prism.DryIoc.Extensions/PrismContainerExtension.cs [hunk 2]
```diff
@@ -155,7 +155,8 @@ public IContainerRegistry RegisterManySingleton(Type implementingType, params Ty
 
         public bool IsRegistered(Type type)
         {
-            return Instance.IsRegistered(type);
+            return Instance.IsRegistered(type) ||
+                Instance.IsRegistered(type, factoryType: FactoryType.Wrapper);
```
Comment (dahlbyk) [inline src/Prism.DryIoc.Extensions/PrismContainerExtension.cs:159]: @dadhi is there a better way to get DryIoc to return `true` when querying for `IEnumerable<T>` (or `T[]` or....)?

Comment (dansiegel) [reply src/Prism.DryIoc.Extensions/PrismContainerExtension.cs:159]: let's keep the conversation going in #124

**Pull request closed without merging**